src/signals-unix.c

```diff
@@ -46,17 +46,101 @@ static bt_context_t *jl_to_bt_context(void *sigctx)
+{
+    // Modifying the ucontext should work but there is concern that
+    // sigreturn orientated programming mitigation can work against us
+    // by rejecting ucontext that is modified.
+    // The current (staged) implementation in the Linux Kernel only
+    // checks that the syscall is made in the signal handler and that
+    // the ucontext address is valid. Hopefully the value of the ucontext
+    // will not be part of the validation...
+    uintptr_t rsp = jl_get_rsp_from_ctx(_ctx);
+    if (!rsp) {
+        fptr();
+        return;
+    }
+    // Align the stack and leave 128bytes red zone on all platforms.
+    // (Only a few actually needs these but doesn't hurt to do on all of them).
+    rsp &= -16; // ensure 16-byte alignment
+    rsp -= 128; // 128bytes red zone
+#if defined(_OS_LINUX_) && defined(_CPU_X86_64_)
+    ucontext_t *ctx = (ucontext_t*)_ctx;
+    rsp -= sizeof(void*);
+    *(void**)rsp = NULL;
+    ctx->uc_mcontext.gregs[REG_RSP] = rsp;
+    ctx->uc_mcontext.gregs[REG_RIP] = (uintptr_t)fptr;
+#elif defined(_OS_LINUX_) && defined(_CPU_X86_)
+    ucontext_t *ctx = (ucontext_t*)_ctx;
+    rsp -= sizeof(void*);
+    *(void**)rsp = NULL;
+    ctx->uc_mcontext.gregs[REG_ESP] = rsp;
+    ctx->uc_mcontext.gregs[REG_EIP] = (uintptr_t)fptr;
+#elif defined(_OS_LINUX_) && defined(_CPU_AARCH64_)
+    ucontext_t *ctx = (ucontext_t*)_ctx;
+    ctx->uc_mcontext.sp = rsp;
+    ctx->uc_mcontext.regs[29] = 0; // Clear link register (x29)
+    ctx->uc_mcontext.pc = (uintptr_t)fptr;
+#elif defined(_OS_LINUX_) && defined(_CPU_ARM_)
+    ucontext_t *ctx = (ucontext_t*)_ctx;
+    ctx->uc_mcontext.arm_sp = rsp;
+    ctx->uc_mcontext.arm_lr = 0; // Clear link register
+    ctx->uc_mcontext.arm_pc = (uintptr_t)fptr;
+#elif defined(_OS_DARWIN_)
+    // Only used for SIGFPE.
+    // This doesn't seems to be reliable when the SIGFPE is generated
+    // from a divide-by-zero exception, which is now handled by
+    // `catch_exception_raise`. It works fine when a signal is recieved
+    // due to `kill`/`raise` though.
+    ucontext64_t *ctx = (ucontext64_t*)_ctx;
+    rsp -= sizeof(void*);
+    *(void**)rsp = NULL;
+    ctx->uc_mcontext64->__ss.__rsp = rsp;
+    ctx->uc_mcontext64->__ss.__rip = (uintptr_t)fptr;
+#else
+    fptr();
+#endif
+}
+
+static void jl_throw_in_ctx(jl_value_t *e, void *sigctx)
 {
     jl_ptls_t ptls = jl_get_ptls_states();
     if (!ptls->safe_restore)
         ptls->bt_size = rec_backtrace_ctx(ptls->bt_data, JL_MAX_BT_SIZE,
                                           jl_to_bt_context(sigctx));
     ptls->exception_in_transit = e;
-    // TODO throw the error by modifying sigctx for supported platforms
-    // This will avoid running the atexit handler on the signal stack
-    // if no excepiton handler is registered.
-    jl_rethrow();
+    // Do not use the main stack if this is a stack overflow since that will
+    // not work....
+    if (e == jl_stackovf_exception)
+        jl_rethrow();
```

Comment:
on `mach`, we re-use the signal stack for the rethrow call. should we eventually do that here too?